engines/twp/twp.cpp

```diff
@@ -0,0 +1,1746 @@
+/* ScummVM - Graphic Adventure Engine
+ *
+ * ScummVM is the legal property of its developers, whose names
+ * are too numerous to list here. Please refer to the COPYRIGHT
+ * file distributed with this source distribution.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#if defined(HAVE_CONFIG_H)
+#include "config.h"
+#endif
+
+#ifdef USE_IMGUI
+#define FORBIDDEN_SYMBOL_ALLOW_ALL
+#include "graphics/imgui/imgui.h"
+#include "graphics/imgui/backends/imgui_impl_sdl2_scummvm.h"
+#include "graphics/imgui/backends/imgui_impl_opengl3_scummvm.h"
+#undef FORBIDDEN_SYMBOL_ALLOW_ALL
+#include "backends/graphics3d/openglsdl/openglsdl-graphics3d.h"
+// here I undefined these symbols because of the <X11/Xlib.h> defining them
+// and messing with all classes or structure using the same names
+#undef Bool
+#undef CursorShape
+#undef Expose
+#undef KeyPress
+#undef KeyRelease
+#undef FocusIn
+#undef FocusOut
+#undef FontChange
+#undef None
+#undef Status
+#undef Unsorted
+
+#endif
+
+#include "common/config-manager.h"
+#include "common/events.h"
+#include "common/savefile.h"
+#include "image/png.h"
+#include "engines/util.h"
+#include "graphics/opengl/system_headers.h"
+
+#include "twp/twp.h"
+#include "twp/console.h"
+#include "twp/lighting.h"
+#include "twp/thread.h"
+#include "twp/squtil.h"
+#include "twp/task.h"
+#include "twp/enginedialogtarget.h"
+#include "twp/actions.h"
+#include "twp/debugtools.h"
+
+namespace Twp {
+
+TwpEngine *g_twp;
+
+#ifdef USE_IMGUI
+SDL_Window *g_window = nullptr;
```

Comment:
Not portable, incorrect. You may have no SDL on the target system. If you need to expose the ScummVM window, then we must add it to OSystem.